core/shared/src/main/scala/magnolia.scala

```diff
@@ -81,32 +81,53 @@ object Magnolia {
     *  parameters apply equally to `dispatch` as to `combine`.
     *  */
   def gen[T: c.WeakTypeTag](c: whitebox.Context): c.Tree = Stack.withContext(c) { stack =>
-    import c.universe._
     import c.internal._
+    import c.universe._
+    import definitions._
+
+    val genericType = weakTypeOf[T]
+    val genericSymbol = genericType.typeSymbol
 
     def error(message: String): Nothing = c.abort(c.enclosingPosition, s"magnolia: $message")
     def warning(message: String): Unit = c.warning(c.enclosingPosition, s"magnolia: $message")
 
-    val debug = c.macroApplication.symbol.annotations
-      .find(_.tree.tpe <:< typeOf[debug])
-      .flatMap(_.tree.children.tail.collectFirst {
-        case Literal(Constant(s: String)) => s
-        case tree if typeOf[debug].companion.decls.exists(_ == tree.symbol) => ""  // Default constructor, i.e. @debug or @debug()
-        case other => error(s"Invalid argument $other in @debug annotation. Only string literals or empty constructor supported")
-      })
-
-    val magnoliaPkg = c.mirror.staticPackage("magnolia")
-    val scalaPkg = c.mirror.staticPackage("scala")
-
-    val repeatedParamClass = definitions.RepeatedParamClass
-    val scalaSeqType = typeOf[Seq[_]].typeConstructor
-    val javaAnnotationType = typeOf[java.lang.annotation.Annotation]
-    val typeClassName = TypeName("Typeclass")
+    val ArrayObj = reify(Array).tree
+    val CallByNeedObj = reify(CallByNeed).tree
+    val CaseClassSym = symbolOf[CaseClass[Any, Any]]
+    val DebugTpe = typeOf[debug]
+    val EitherSym = symbolOf[Either[Any, Any]]
+    val JavaAnnotationTpe = typeOf[java.lang.annotation.Annotation]
+    val LeftObj = reify(Left).tree
+    val MagnoliaUtilObj = reify(MagnoliaUtil).tree
+    val MercatorOpsSym = symbolOf[mercator.Ops[Any, Any]]
+    val MonadicSym = symbolOf[mercator.Monadic[Any]]
+    val NoneObj = reify(None).tree
+    val ParamObj = reify(Param).tree
+    val ParamSym = symbolOf[Param[Any, Any]]
+    val ReadOnlyCaseClassSym = symbolOf[ReadOnlyCaseClass[Any, Any]]
+    val ReadOnlyParamObj = reify(ReadOnlyParam).tree
+    val ReadOnlyParamSym = symbolOf[ReadOnlyParam[Any, Any]]
+    val RightObj = reify(Right).tree
+    val SealedTraitSym = symbolOf[SealedTrait[Any, Any]]
+    val SeqTpe = typeOf[Seq[Any]].typeConstructor
+    val SomeObj = reify(Some).tree
+    val SubtypeObj = reify(Subtype).tree
+    val SubtypeTpe = typeOf[Subtype[Any, Any]].typeConstructor
+    val TypeClassNme = TypeName("Typeclass")
+    val TypeNameObj = reify(magnolia.TypeName).tree
```

Comment:
This looks much cleaner. Do you think there's any performance benefit to making them lazy?

Reply:
I was wondering the same. The cost is wrapping them in `LazyRef`. Do you think we should?

Reply:
Let's make it a future enhancement. ;)